vault/raft.go

```diff
@@ -0,0 +1,664 @@
+package vault
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	proto "github.com/golang/protobuf/proto"
+	"github.com/hashicorp/errwrap"
+	cleanhttp "github.com/hashicorp/go-cleanhttp"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/physical/raft"
+	"github.com/hashicorp/vault/sdk/helper/jsonutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/sdk/physical"
+	"github.com/hashicorp/vault/vault/seal"
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/net/http2"
+)
+
+var (
+	raftTLSStoragePath    = "core/raft/tls"
+	raftTLSRotationPeriod = 24 * time.Hour
+)
+
+type raftFollowerStates struct {
+	l         sync.RWMutex
+	followers map[string]uint64
+}
+
+func (s *raftFollowerStates) update(nodeID string, appliedIndex uint64) {
+	s.l.Lock()
+	s.followers[nodeID] = appliedIndex
+	s.l.Unlock()
+}
+func (s *raftFollowerStates) delete(nodeID string) {
+	s.l.RLock()
+	delete(s.followers, nodeID)
+	s.l.RUnlock()
+}
+func (s *raftFollowerStates) get(nodeID string) uint64 {
+	s.l.RLock()
+	index := s.followers[nodeID]
+	s.l.RUnlock()
+	return index
+}
+func (s *raftFollowerStates) minIndex() uint64 {
+	var min uint64 = math.MaxUint64
+	minFunc := func(a, b uint64) uint64 {
+		if a > b {
+			return b
+		}
+		return a
+	}
+
+	s.l.RLock()
+	for _, i := range s.followers {
+		min = minFunc(min, i)
+	}
+	s.l.RUnlock()
+
+	if min == math.MaxUint64 {
+		return 0
+	}
+
+	return min
+}
+
+func (c *Core) startRaftStorage(ctx context.Context) error {
+	if raftStorage, ok := c.underlyingPhysical.(*raft.RaftBackend); ok {
```

Comment:
As a nicety, can you invert this so we just exit immediately if `!ok`. That way the entirety of the function isn't in an if block.